### PR TITLE
server: Add GetSecret, GetSecrets and SetSecret implementation

### PR DESCRIPTION
**Pull request by warusadura**
This change adds Secret Service method GetSecrets and Secret Item methods GetSecret and SetSecret.

### CODE REVIEW

File: server/src/item.rs [hunk 1]
```diff
@@ -32,8 +32,32 @@ impl Item {
...
+        session: OwnedObjectPath,
+    ) -> Result<(SecretInner,), ServiceError> {
+        // wip
+        let inner = self.inner.lock().await;
+        let manager = self.manager.lock().await;
+        let secret = inner.secret();
+        let session = manager.session(&session);
+        if session.is_none() {
+            tracing::error!("The session does not exist");
+            return Err(ServiceError::NoSession("NoSession".to_string()));
```
Comment (bilelmoussaoui) [inline server/src/item.rs:46]: It is useless to return NoSession in an error message that is already named NoSession. Instead pass the info about which session or a more expressive error message :)

File: server/src/item.rs [hunk 2]
```diff
@@ -32,8 +32,32 @@ impl Item {
...
+        }
+
+        let session = session.unwrap();
```
Comment (bilelmoussaoui) [inline server/src/item.rs:49]: Use let Some(session) = manager.session(&session) else { // the error bits } and you won't need the unwrap here.

File: server/src/item.rs [hunk 3]
```diff
@@ -32,8 +32,32 @@ impl Item {
...
+        let iv = oo7::crypto::generate_iv();
+        let key = session.aes_key();
+
+        let secret = oo7::crypto::encrypt(secret, key, iv.clone());
```
Comment (bilelmoussaoui) [inline server/src/item.rs:53]:
```suggestion
        let secret = oo7::crypto::encrypt(secret, key, &iv);
```
no?

File: server/src/service_manager.rs [hunk 4]
```diff
@@ -29,6 +29,10 @@ impl ServiceManager {
         self.sessions.len()
     }
 
+    pub fn session(&self, path: &OwnedObjectPath) -> Option<Arc<Session>> {
+        Some(Arc::clone(self.sessions.get(path)?))
```
Comment (bilelmoussaoui) [inline server/src/service_manager.rs:33]: What. Use map please :)

File: server/src/session.rs [hunk 5]
```diff
@@ -34,12 +34,16 @@ impl Session {
...
             manager,
         }
     }
 
     pub fn path(&self) -> &OwnedObjectPath {
         &self.path
     }
+
+    pub fn aes_key(&self) -> &Arc<Key> {
+        self.aes_key.as_ref().unwrap()
```
Comment (bilelmoussaoui) [inline server/src/session.rs:47]: This is wrong. A "plain" session would crash here.

File: server/src/item.rs [hunk 6]
```diff
@@ -32,8 +32,32 @@ impl Item {
...
-    pub async fn get_secret(&self, _session: ObjectPath<'_>) -> Result<SecretInner, ServiceError> {
-        todo!()
+    pub async fn get_secret(
+        &self,
+        session: OwnedObjectPath,
+    ) -> Result<(SecretInner,), ServiceError> {
+        // wip
+        let inner = self.inner.lock().await;
+        let manager = self.manager.lock().await;
+        let secret = inner.secret();
```
Comment (bilelmoussaoui) [inline server/src/item.rs:42]: Move this after getting the session, no need to "read" the secret if you will fail later anyways

File: server/src/item.rs (hunk 3 above)
Comment (bilelmoussaoui) [inline server/src/item.rs:53]: There is nothing to encrypt if there is no aes_key

File: server/src/item.rs (hunk 1 above)
Comment (warusadura) [reply server/src/item.rs:46]: Thanks for the review. Yeah of course. This was just placed temporarily.

File: server/src/item.rs (hunk 3 above)
Comment (warusadura) [reply server/src/item.rs:53]: Yes.

File: server/src/session.rs (hunk 5 above)
Comment (warusadura) [reply server/src/session.rs:47]: Yeah, I missed that. Thanks!

File: server/src/item.rs (hunk 3 above)
Comment (warusadura) [reply server/src/item.rs:53]: Yeah, I agree.

File: server/src/service_manager.rs (hunk 4 above)
Comment (warusadura) [reply server/src/service_manager.rs]: yeah, done :)

File: server/src/item.rs (hunk 6 above)
Comment (warusadura) [reply server/src/item.rs]: Done, thanks!

File: server/src/item.rs (hunk 2 above)
Comment (warusadura) [reply server/src/item.rs]: Done, thanks!

File: server/src/session.rs [hunk 7]
```diff
@@ -34,12 +34,20 @@ impl Session {
...
         &self.path
     }
+
+    pub fn aes_key(&self) -> Option<&Key> {
+        if let Some(key) = &self.aes_key {
+            Some(key.as_ref())
+        } else {
+            None
+        }
+    }
```
Comment (bilelmoussaoui) [inline server/src/session.rs:52]: Return a Option<Arc<Key>> like you did for session.

File: server/src/service.rs [hunk 8]
```diff
@@ -235,7 +263,6 @@ impl Service {
             Arc::clone(&service.manager),
             Arc::new(Keyring::temporary(Secret::random()).await?),
         );
-        collections.push(collection.clone());
```
Comment (bilelmoussaoui) [inline server/src/service.rs:238]: why was this removed?

File: server/src/service.rs [hunk 9]
```diff
@@ -107,10 +107,38 @@ impl Service {
...
+        let mut secrets: HashMap<OwnedObjectPath, SecretInner> = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.item_inner(item).await {
+                    match item.get_secret(session.clone()).await {
+                        Ok(secret) => {
+                            secrets.insert(item.path().clone(), secret.0);
+                            if secrets.len() == items.len() {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:122]: What is this condition for?

File: server/src/service.rs [hunk 10]
```diff
@@ -107,10 +107,38 @@ impl Service {
...
+                                break 'outer;
+                            }
+                        }
+                        Err(ServiceError::NoSession(err)) => {
+                            return Err(ServiceError::NoSession(err));
+                        }
+                        Err(_) => (),
```
Comment (bilelmoussaoui) [inline server/src/service.rs:129]: This sounds just wrong?

File: server/src/service.rs [hunk 11]
```diff
@@ -107,10 +107,38 @@ impl Service {
...
+        items: Vec<OwnedObjectPath>,
+        session: OwnedObjectPath,
     ) -> Result<HashMap<OwnedObjectPath, SecretInner>, ServiceError> {
-        todo!()
+        let mut secrets: HashMap<OwnedObjectPath, SecretInner> = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.item_inner(item).await {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:118]: Instead of adding this function, you should make the search_items_inner function from the that other PR takes a search condition function, that way you can use it both for comparing attributes or item paths.

File: server/src/service.rs [hunk 12]
```diff
@@ -107,10 +107,38 @@ impl Service {
     #[zbus(out_args("secrets"))]
     pub async fn get_secrets(
         &self,
-        _items: Vec<OwnedObjectPath>,
-        _session: ObjectPath<'_>,
+        items: Vec<OwnedObjectPath>,
+        session: OwnedObjectPath,
     ) -> Result<HashMap<OwnedObjectPath, SecretInner>, ServiceError> {
-        todo!()
+        let mut secrets: HashMap<OwnedObjectPath, SecretInner> = HashMap::new();
```
Comment (bilelmoussaoui) [inline server/src/service.rs:113]: the type is not needed here.

File: server/src/item.rs [hunk 13]
```diff
@@ -32,8 +32,49 @@ impl Item {
...
+        if self.is_locked().await {
+            tracing::error!("Cannot get secret of a locked object.");
+            return Err(ServiceError::IsLocked(
+                "Cannot get secret of a locked object.".to_string(),
+            ));
+        }
+
+        let inner = self.inner.lock().await;
+        let secret = inner.secret();
+        let iv = oo7::crypto::generate_iv();
```
Comment (bilelmoussaoui) [inline server/src/item.rs:60]: Is there really an iv passed for a not encrypted session?

File: server/src/item.rs [hunk 14]
```diff
@@ -32,8 +32,49 @@ impl Item {
...
+        let session_inner;
+
+        if let Some(session) = manager.session(&session) {
+            session_inner = session;
+        } else {
+            tracing::error!("The session: {} does not exist.", session);
+            return Err(ServiceError::NoSession(
+                "The session does not exist.".to_string(),
+            ));
+        }
```
Comment (bilelmoussaoui) [inline server/src/item.rs:49]:
```suggestion
        let Some(session) = manager.session(&session) else {
            tracing::error!("The session: {} does not exist.", session);
            return Err(ServiceError::NoSession(
                "The session does not exist.".to_string(),
            ));
        }
```

File: server/src/service.rs (hunk 11 above)
Comment (warusadura) [reply server/src/service.rs:118]: Thanks! Sorry for the late reply :)
Yeah it make sense. But, the returns are different. For this case we're returning a single Item based on the objectpath and in SearchItems we're returning multiple Items based on attributes.
So, it is bad to have two methods: `search_inner_items(attributes)` and `inner_item(objectpath)`?

Comment (bilelmoussaoui) [reply server/src/service.rs:118]: The name is just so weird. Maybe find_item_with_path or so?

Comment (warusadura) [reply server/src/service.rs:118]: yeah, I agree :) will update it.

File: server/src/session.rs (hunk 7 above)
Comment (warusadura) [reply server/src/session.rs]: Done, thanks!

File: server/src/item.rs (hunk 13 above)
Comment (warusadura) [reply server/src/item.rs]: Yeah, removed.

File: server/src/service.rs (hunk 12 above)
Comment (warusadura) [reply server/src/service.rs]: Thanks, removed.

File: server/src/item.rs (hunk 14 above)
Comment (warusadura) [reply server/src/item.rs]: Done, thanks!

File: server/src/service.rs (hunk 8 above)
Comment (warusadura) [reply server/src/service.rs:238]: This change is introduced from a separate [commit](https://github.com/bilelmoussaoui/oo7/pull/153/commits/764ddb9f36a4db498caa77cd646cebc346890525).

> Since session collection is a temporary collection without any items
we shouldn't push it to the Service Collections.

With this change we can avoid an extra iteration when looking for items.

File: server/src/service.rs (hunk 9 above)
Comment (warusadura) [reply server/src/service.rs:122]: This is to avoid iterating through all the available collections (Service.Collections). If we have found the matching items to dbus call arguments, there is no need to continue iterating through remaining collections.

File: server/src/service.rs (hunk 10 above)
Comment (warusadura) [reply server/src/service.rs]: Modified to return the error.

File: server/src/service.rs (hunk 8 above)
Comment (warusadura) [reply server/src/service.rs]: Yeah, thanks! Removed.

File: server/src/item.rs [hunk 15]
```diff
@@ -32,8 +32,47 @@ impl Item {
...
+        };
+
+        tracing::info!("Secret retrieved from the item: {}.", self.path);
+
+        Ok((SecretInner(
+            session.path().clone(),
+            iv,
+            encrypted,
+            "text/plain".to_string(),
+        ),))
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/item.rs]: This is much cleaner, thanks!

File: server/src/service.rs [hunk 16]
```diff
@@ -107,10 +107,40 @@ impl Service {
...
+                        }
+                        Err(err) => {
+                            return Err(err);
+                        }
+                    };
+                } else {
+                    tracing::info!(
+                        "Item: {:?} does not exist in collection: {}",
+                        item,
+                        collection.path().clone()
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/service.rs]: Removed.

File: server/src/item.rs [hunk 17]
```diff
@@ -32,8 +32,47 @@ impl Item {
...
+
+        let Some(session) = manager.session(&session) else {
+            tracing::error!("The session: {} does not exist.", session);
+            return Err(ServiceError::NoSession(
+                "The session does not exist.".to_string(),
+            ));
+        };
+
+        if self.is_locked().await {
+            tracing::error!("Cannot get secret of a locked object.");
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/item.rs:50]: Done.

File: server/src/item.rs [hunk 18]
```diff
@@ -32,8 +32,47 @@ impl Item {
...
+    pub async fn get_secret(
+        &self,
+        session: OwnedObjectPath,
+    ) -> Result<(SecretInner,), ServiceError> {
+        let manager = self.manager.lock().await;
+
+        let Some(session) = manager.session(&session) else {
+            tracing::error!("The session: {} does not exist.", session);
+            return Err(ServiceError::NoSession(
+                "The session does not exist.".to_string(),
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/item.rs]: Done, thanks!

File: server/src/collection.rs [hunk 19]
```diff
@@ -157,6 +157,18 @@ impl Collection {
         self.alias.lock().await.clone()
     }
 
+    pub async fn get_item_from_path(&self, path: &OwnedObjectPath) -> Option<item::Item> {
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/collection.rs]: Done, thanks!

File: server/src/service.rs [hunk 20]
```diff
@@ -107,10 +107,40 @@ impl Service {
...
+        session: OwnedObjectPath,
     ) -> Result<HashMap<OwnedObjectPath, SecretInner>, ServiceError> {
-        todo!()
+        let mut secrets = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.get_item_from_path(item).await {
+                    match item.get_secret(session.clone()).await {
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/service.rs:143]: Yeah I could have done this. But, actually we need the match. Because, consider a scenario where a correct Session is present/provided. But the item(s) are locked. In this situation we still need to return `secrets: HashMap<OwnedObjectPath, SecretInner>`. We can error out only if there is no session (returns nothing). Of course this is based on gnome-keyring behavior :)

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply server/src/service.rs:143]: The match you wrote doesn't do what you are mentioning :)

[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/service.rs:143]: Ha :) yes, you're right.

The following is what we need:
Based on gnome-keyring behavior,  `Item.get_secret` call error out if there is no session or the item is locked. But `Service.get_secrets` only error out when there is no session.

I had this correctly implemented before [this](https://github.com/bilelmoussaoui/oo7/pull/153#discussion_r1844014981) change,

I added another arm:
```
Err(ServiceError::IsLocked(_)) => {                                                   
      continue;
 }
```

with this we don't error out anymore if the items are locked when calling Service.get_secrets.

If this is still bad, how can I improve this?

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply server/src/service.rs:143]: Well, first of all this difference in behaviour between Item::get_secret & Service::get_secrets should be documented in Service::get_secrets along with a reference to gnome-keyring-daemon.

Also, you don't need all the other match statements, only the Ok/Err(IsLocked)/Err(e), are enough.

File: server/src/item.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply server/src/item.rs:50]: No, you only did for the error, not the logged message. The error is received by the client, the logs is what we might all have in some scenarios.

File: server/src/service.rs (hunk 9 above)
Comment (bilelmoussaoui) [reply server/src/service.rs:146]: Then add a small comment explaining why.

File: server/src/service.rs [hunk 21]
```diff
@@ -131,10 +131,43 @@ impl Service {
...
+                            return Err(err);
+                        }
+                    };
+                } else {
+                    tracing::debug!(
+                        "Item: {:?} does not exist in collection: {}",
+                        item,
+                        collection.path()
+                    );
+                }
```
Comment (bilelmoussaoui) [inline server/src/service.rs:166]: This is useless.

Comment (bilelmoussaoui) [reply server/src/service.rs:166]: I meant the debugging message, it will just print too many debug logs for nothing

Comment (warusadura) [reply server/src/service.rs]: Removed.

File: server/src/service.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/service.rs:143]: Got it, thanks!

File: server/src/service.rs (hunk 9 above)
Comment (warusadura) [reply server/src/service.rs:148]: Done.

File: server/src/service.rs [hunk 22]
```diff
@@ -131,10 +131,43 @@ impl Service {
...
-        _items: Vec<OwnedObjectPath>,
-        _session: ObjectPath<'_>,
+        items: Vec<OwnedObjectPath>,
+        session: OwnedObjectPath,
     ) -> Result<HashMap<OwnedObjectPath, SecretInner>, ServiceError> {
-        todo!()
+        let mut secrets = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
```
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/service.rs:140]: Why though :) is it bad practice to use a labeled loop block?

File: server/src/item.rs [hunk 23]
```diff
@@ -32,12 +32,83 @@ impl Item {
...
+                session
+            )));
+        };
+
+        let mut inner = self.inner.lock().await;
+
+        match session.aes_key() {
+            Some(key) => {
+                let iv = secret.1;
+                let secret_encrypted = secret.2;
```
Comment (bilelmoussaoui) [inline server/src/item.rs:100]:
```suggestion
                let {session, iv, encrypted_secret, _content_type} = secret;
```
move it to the top

Comment (bilelmoussaoui) [reply server/src/item.rs:100]: we will handle content_type later on in the future.

File: server/src/service.rs [hunk 24]
```diff
@@ -131,10 +131,38 @@ impl Service {
...
+                        Ok(secret) => {
+                            secrets.insert(item.path().clone(), secret.0);
+                            // To avoid iterating through all the remaining collection, if the items
+                            // secrets are already retrieved.
+                            if secrets.len() == items.len() {
+                                break 'outer;
+                            }
+                        }
+                        // Based on gnome-keyring-daemon behavior, to avoid error out when the item
+                        // is Locked.
```
Comment (bilelmoussaoui) [inline server/src/service.rs:153]:
```suggestion
                        // Avoid erroring out if an item is locked.
```

File: server/src/service.rs [hunk 25]
```diff
@@ -131,10 +131,38 @@ impl Service {
...
+        let mut secrets = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.item_from_path(item).await {
+                    match item.get_secret(session.clone()).await {
+                        Ok(secret) => {
+                            secrets.insert(item.path().clone(), secret.0);
+                            // To avoid iterating through all the remaining collection, if the items
```
Comment (bilelmoussaoui) [inline server/src/service.rs:146]:
```suggestion
                            // To avoid iterating through all the remaining collections, if the items
```

File: server/src/service.rs [hunk 26]
```diff
@@ -131,10 +131,38 @@ impl Service {
...
-        todo!()
+        let mut secrets = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.item_from_path(item).await {
+                    match item.get_secret(session.clone()).await {
+                        Ok(secret) => {
+                            secrets.insert(item.path().clone(), secret.0);
```
Comment (bilelmoussaoui) [inline server/src/service.rs:145]:
```suggestion
                        Ok((secret,)) => {
                            secrets.insert(item.path().clone(), secret);
```

File: server/src/service.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply server/src/service.rs:140]: Nothing per see, just trying to understand what is the most efficient way to do all of this.

**Review by bilelmoussaoui — COMMENTED** [review]
almost there

**Comment by bilelmoussaoui** [comment]
The commit message and it description could use some work, in general. Currently you just duplicate the commit message "title" with the "description". In the description, add any technical information or detail or leave it empty.

File: server/src/item.rs (hunk 23 above)
Comment (warusadura) [reply server/src/item.rs]: Done, thanks!

File: server/src/service.rs (hunk 24 above)
Comment (warusadura) [reply server/src/service.rs]: Done.

File: server/src/service.rs (hunk 25 above)
Comment (warusadura) [reply server/src/service.rs]: Done.

File: server/src/service.rs (hunk 26 above)
Comment (warusadura) [reply server/src/service.rs]: Done, thanks!
